pkg/controller/certificates/certificate_controller.go

```diff
@@ -186,15 +179,21 @@ func (cc *CertificateController) maybeSignCertificate(key string) error {
 	defer func() {
 		glog.V(4).Infof("Finished syncing certificate request %q (%v)", key, time.Now().Sub(startTime))
 	}()
-	obj, exists, err := cc.csrStore.Store.GetByKey(key)
+	csr, err := cc.csrLister.Get(key)
+	if errors.IsNotFound(err) {
+		glog.V(3).Infof("csr has been deleted: %v", key)
+		return nil
+	}
 	if err != nil {
 		return err
 	}
-	if !exists {
-		glog.V(3).Infof("csr has been deleted: %v", key)
-		return nil
+
+	// need to operate on a copy so we don't mutate the csr in the shared cache
+	csrCopy, err := api.Scheme.DeepCopy(csr)
```

Comment:
move the copy (maybe need a utility function to avoid ugly) to the inside of the `if` blocks or make a short-circuit above your copy.  there's no need to copy if we aren't going to be doing work (the common case).